### PR TITLE
Offer Reset: always displays owned products regardless of the selected term

### PR DESCRIPTION
**Pull request by rcanepa**
#### Changes proposed in this Pull Request

* Follow the pattern already established in the previous version of the Plans grid. That pattern says that the owned plan and products should always be visible, and highlighted as such, regardless of the selected term. This means that, for example, users should not be able to purchase Jetpack Backup Daily monthly when they already own the yearly version. I our current state, they can.
* Offer a path to upgrade a monthly plan subscription to a yearly one.

#### Testing instructions

Try to pay attention to duplicate cards, label of call to actions, and badges. Everything should make sense. When testing, make sure to purchase both monthly and yearly subscriptions.

There are several cases we should cover to make sure we got all right.

* Individual purchases:
  * Plans: purchase a plan and make sure the card of the plan is carried over regardless the selected term. Make sure products included in the plan behave the same.
  * Products with option: purchase a product, such as Jetpack Backup Daily, and make sure that the card of the product is carried over regardless the selected term.
  * Products without option: purchase a product, such as Jetpack Search, and make sure that the card of the product is carried over regardless the selected term.

* Multiple purchases:
   * Plan + product: purchase a plan and a product. An example could be Jetpack Security Real-time and Jetpack Search. Verify that the cards of the plan and product are carried over regardless the selected term.

* Upgrade to yearly:
   * Purchase a monthly subscription to any plan and selected the Yearly term from the filter bar.
   * Verify that the CTA label says: "Upgrade to Yearly".

Fixes 1169247016322522-as-1192817246446001

###### Monthly
image
###### Yearly
image

Jetpack Security
###### Monthly
image
###### Yearly (Upgrade to Yearly)
image

### CODE REVIEW

File: client/my-sites/plans-v2/plans-column/index.tsx [hunk 1]
```diff
@@ -33,6 +37,8 @@ const PlansColumn = ( { duration, onPlanClick, productType, siteId }: PlanColumn
 	const currentPlan =
 		useSelector( ( state ) => getSitePlan( state, siteId ) )?.product_slug || null;
 
+	const optionsFromCurrentPlan = getAllOptionsFromSlug( currentPlan ) || [];
```
Comment (rcanepa) [inline client/my-sites/plans-v2/plans-column/index.tsx:40]: If a user owns `jetpack_backup_daily_monthly`, we don't want to display the Jetpack Backup option card, therefore, we search for all slugs of the Jetpack Backup option cards (there are two, a monthly and a yearly).

File: client/my-sites/plans-v2/product-card/index.tsx [hunk 2]
```diff
@@ -165,16 +171,16 @@ const ProductCardWrapper = ( {
...
-			onButtonClick={ () => onClick( item, purchase ) }
+			onButtonClick={ () => onClick( item, isUpgradeableToYearly, purchase ) }
 			features={ item.features }
 			children={ item.children }
 			originalPrice={ originalPrice }
 			discountedPrice={ discountedPrice }
 			withStartingPrice={
 				// Search has several pricing tiers
-				item.subtypes.length > 0 || JETPACK_SEARCH_PRODUCTS.includes( item.product_slug )
+				item.subtypes.length > 0 || JETPACK_SEARCH_PRODUCTS.includes( item.productSlug )
```
Comment (rcanepa) [inline client/my-sites/plans-v2/product-card/index.tsx:183]: This seems to be a bug that went unnoticed.

File: client/my-sites/plans-v2/product-card/index.tsx [hunk 3]
```diff
@@ -186,7 +192,7 @@ const ProductCardWrapper = ( {
 			hidePrice={
 				// Don't hide price if siteId is not defined, since it most likely won't be shown
 				// in other parts of the card (e.g. Jetpack Search)
-				siteId && item.hidePrice
+				!! siteId && item.hidePrice
```
Comment (rcanepa) [inline client/my-sites/plans-v2/product-card/index.tsx:195]: TypeScript fix.

File: client/my-sites/plans-v2/products-column/index.tsx [hunk 4]
```diff
@@ -56,30 +56,46 @@ const ProductsColumn = ( {
...
-				} ) ),
-		[ duration, includedInPlanProducts, ownedProducts, productType ]
-	);
+	const productObjects: SelectorProduct[] = useMemo( () => {
+		// Owned products (plans are filtered out).
+		const ownedProducts = siteProducts
+			.map( ( { productSlug } ) => productSlug )
+			.filter( ( productSlug ) => JETPACK_PRODUCTS_LIST.includes( productSlug ) )
+			.map( slugToSelectorProduct )
+			.filter( ( product ): product is SelectorProduct => product !== null );
```
Comment (rcanepa) [inline client/my-sites/plans-v2/products-column/index.tsx:65]: Now owned products are always displayed. They don't go through the filtering process the rest of the cards do.

**Comment by monsieur-z** [comment]
I've bought Complete monthly. When I visit the Plans page and select the yearly term, I see Complete twice.
Screen Shot 2020-09-09 at 2 43 19 PM
Screen Shot 2020-09-09 at 2 43 27 PM

**Comment by monsieur-z** [comment]
I purchased Anti-spam yearly. Same as above, when I visit the Plans page with the monthly term, I see the card twice.

File: client/my-sites/plans-v2/plans-column/index.tsx (hunk 1 above)
Comment (monsieur-z) [reply client/my-sites/plans-v2/plans-column/index.tsx:40]: There's a small TS error at this line.

File: client/my-sites/plans-v2/plans-column/index.tsx [hunk 5]
```diff
@@ -43,10 +49,10 @@ const PlansColumn = ( { duration, onPlanClick, productType, siteId }: PlanColumn
 					!! product &&
 					duration === product.term &&
 					PRODUCTS_TYPES[ productType ].includes( product.productSlug ) &&
-					// Don't include a generic/option card if the user already owns a subtype
-					! product.subtypes.includes( currentPlan || '' ) &&
 					// Don't include a card of a plan the user already owns
-					product.productSlug !== currentPlan
+					product.productSlug !== currentPlan &&
+					// Don't include a generic/option card if the user already owns a subtype
+					! optionsFromCurrentPlan.includes( product.productSlug )
```
Comment (monsieur-z) [inline client/my-sites/plans-v2/plans-column/index.tsx:55]: `optionsFromCurrentPlan` should be added to the list of deps for `useMemo`.

**Review by monsieur-z — CHANGES_REQUESTED** [review]
The only weird behaviour I spotted is a duplicated card when you own a term and the other one is shown. Other that, there are 2 minor TS errors.

File: client/my-sites/plans-v2/plans-column/index.tsx (hunk 1 above)
Comment (rcanepa) [reply client/my-sites/plans-v2/plans-column/index.tsx:40]: Good catch!

File: client/my-sites/plans-v2/plans-column/index.tsx (hunk 5 above)
Comment (rcanepa) [reply client/my-sites/plans-v2/plans-column/index.tsx:55]: Thanks!

**Comment by rcanepa** [comment]
> I've bought Complete monthly. When I visit the Plans page and select the yearly term, I see Complete twice.

Thanks for giving it such a deep review @monsieur-z. I have updated the PR to address all issues, I hope. I created a hook to manage what product cards should be displayed in the Plans grid. Hopefully, it will bring more clarity to the logic behind it.

**Review by wp-desktop — CHANGES_REQUESTED** [review]
WordPress Desktop CI Failure for job "wp-desktop-mac".

@rcanepa please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/882368). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

**Comment by monsieur-z** [comment]
> > I've bought Complete monthly. When I visit the Plans page and select the yearly term, I see Complete twice.
> 
> Thanks for giving it such a deep review @monsieur-z. I have updated the PR to address all issues, I hope. I created a hook to manage what product cards should be displayed in the Plans grid. Hopefully, it will bring more clarity to the logic behind it.

Love the hook @rcanepa!

**Review by monsieur-z — APPROVED** [review]
Everything seems to work as expected. Good job!

**Review by elliottprogrammer — APPROVED** [review]
I tested with various combinations of yearly and monthly plans and products. Works excellent as expected.  Code looks great. Very nice abstraction with `useGetPlansGridProducts()`!  Nice work @rcanepa ! 👍

**Pull request merged**